docs/develop/typescript/message-passing.mdx

```diff
@@ -196,107 +290,319 @@ await client.workflow.signalWithStart(yourWorkflow, {
+// Workflow code, and to send Updates from Client code.
+export const setLanguageUsingActivity = wf.defineUpdate<Language, [Language]>('setLanguageUsingActivity');
+
+export async function greetingWorkflow(): Promise<string> {
+  const greetings: Partial<Record<Language, string>> = {
+    [Language.CHINESE]: '你好，世界',
+    [Language.ENGLISH]: 'Hello, world',
+  };
+
+  let language = Language.ENGLISH;
+
+  const lock = new Mutex();
+  wf.setHandler(setLanguageUsingActivity, async (newLanguage) => {
+    // 👉 An Update handler can mutate the Workflow state and return a value.
+    // 👉 Since this update handler is async, it can execute an activity.
+    if (!(newLanguage in greetings)) {
+      // 👉 Do the following with the lock held to ensure that multiple calls to set_language are processed in order.
+      await lock.runExclusive(async () => {
+        // Simulate an async call to a greeting service
+        await new Promise((resolve) => setTimeout(resolve, 200));
+        if (!(newLanguage in greetings)) {
+          const greeting = await callGreetingService(newLanguage);
+          if (!greeting) {
+            // 👉 An update validator cannot be async, so cannot be used to check that the remote
+            // call_greeting_service supports the requested language. Raising ApplicationError
+            // will fail the Update, but the WorkflowExecutionUpdateAccepted event will still be
+            // added to history.
+            throw new wf.ApplicationFailure(`${newLanguage} is not supported by the greeting service`);
+          }
+          greetings[newLanguage] = greeting;
+        }
+      });
+    }
+    const previousLanguage = language;
+    language = newLanguage;
+    return previousLanguage;
+  });
+  ...
 }
 ```
 
-<!--SNIPEND-->
+After updating the code to use `async`, your Update handler can schedule an Activity and await the result.
+Although an `async` Signal handler can also execute an Activity, using an Update handler allows the client to receive a result or error once the Activity completes.
+This lets your client track the progress of asynchronous work performed by the Update's Activities, Child Workflows, etc.
 
-### Send a Query {#send-query}
+### Add wait conditions to block
 
-Queries are sent from a Temporal Client.
+Sometimes, `async` Signal or Update handlers need to meet certain conditions before they should continue.
+You can use [`workflow.condition`](https://typescript.temporal.io/api/namespaces/workflow#condition) to prevent the code from proceeding until a condition is true.
+You specify the condition by passing a function that returns `true` or `false`.
+This is an important feature that helps you control your handler logic.
 
-Use [`WorkflowHandle.query`](https://typescript.temporal.io/api/interfaces/client.WorkflowHandle/#query) to query a running or completed Workflow.
+Here are two important use cases for `workflow.condition`:
```

Comment:
We should also mention the very common use case of "wait for a signal (or update) to come in".

```
  let gotSignal = false;
  setHandler(defineSignal(signalName), () => { gotSignal = true; });
  await condition(() => gotSignal, timeout);
```

Reply:
Good call! Added to Typescript and Python.